### PR TITLE
Ditto Kinesis

### PR DESCRIPTION
**Pull request by ben-foster-cko**
This PR adds a new version of Ditto that supports writing to Kinesis.

### CODE REVIEW

File: src/Ditto.Core/SourceSettings.cs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+namespace Ditto.Core
+{
+    public class SourceSettings
```
Comment (joe-hutchinson-cko) [inline src/Ditto.Core/SourceSettings.cs]: Plans for this guy?

File: src/Ditto/ReplicatingConsumerFactory.cs [hunk 2]
```diff
@@ -6,11 +6,11 @@ namespace Ditto
 {
     public class ReplicatingConsumerFactory
     {
-        private readonly AppSettings _appSettings;
+        private readonly DittoSettings _appSettings;
         private readonly ILogger _logger;
         private Lazy<IEventStoreConnection> _destinationConnection;
```
Comment (joe-hutchinson-cko) [inline src/Ditto/ReplicatingConsumerFactory.cs]: I think this can be readonly

File: src/Ditto.Core/SourceSettings.cs [hunk 3]
```diff
@@ -0,0 +1,7 @@
+namespace Ditto.Core
```
Comment (joe-hutchinson-cko) [inline src/Ditto.Core/SourceSettings.cs]: Other namespaces in the project need updating to `Ditto.Core`

File: src/Ditto.Kinesis/appsettings.Development.json [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+          "restrictedToMinimumLevel": "Debug"
+        }
+      }
+    ],
+    "Enrich": [
+      "FromLogContext", 
+      "WithMachineName"
+    ],    
+    "Properties": {
+      "ApplicationName": "Ditto"
```
Comment (joe-hutchinson-cko) [inline src/Ditto.Kinesis/appsettings.Development.json]: `Ditto.Kinesis`

File: src/Ditto.Kinesis/ReplicatingConsumer.cs [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+            };
+
+            using (_logger.OperationAt(LogEventLevel.Debug).Time("Replicating {EventType} #{EventNumber} from {StreamName} (Original Event: #{OriginalEventNumber}) to Kinesis",
+                resolvedEvent.Event.EventType,
+                resolvedEvent.Event.EventNumber,
+                resolvedEvent.Event.EventStreamId,
+                resolvedEvent.OriginalEventNumber))
+            using (DittoMetrics.IODuration.WithIOLabels("kinesis", _kinesisSettings.StreamName, "put_record").NewTimer())
+            {
+                await _kinesis.PutRecordAsync(request);
```
Comment (joe-hutchinson-cko) [inline src/Ditto.Kinesis/ReplicatingConsumer.cs:72]: We ideally want to batch write multiple events.

Comment (joe-hutchinson-cko) [reply src/Ditto.Kinesis/ReplicatingConsumer.cs:72]: Is there any retry strategies here? You'll see two exceptions in normal operation, `ProvisionedThroughputExceededException` and occasionally `AmazonKinesisException` when something goes wrong in AWS systems. Typically I catch all exceptions around put and just do a simple Polly exponential backoff with jitter, could be useful in the case of having many Dittos.

**Review by joe-hutchinson-cko — COMMENTED** [review]
Local running instructions results in the following exception.
image

image

I checked container logs but no errors as far as I can see.

File: src/Ditto.Core/SourceSettings.cs (hunk 1 above)
Comment (ben-foster-cko) [reply src/Ditto.Core/SourceSettings.cs]: None. Removed.

File: docker-setup/localstack/init.sh [hunk 6]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
```
Comment (joe-hutchinson-cko) [inline docker-setup/localstack/init.sh:1]: We've got some whitespace issue.

`ditto-localstack     | /docker-entrypoint-initaws.d/init.sh: line 2: $'\r': command not found`

This works okay. 
```
#!/bin/bash
aws configure set aws_access_key_id "XXX"
aws configure set aws_secret_access_key	"XXX"
aws configure set aws_session_token "XXX"
aws configure set region "eu-west-2"
aws configure set output "json"
aws --endpoint-url=http://localstack:4566 kinesis create-stream --stream-name ditto --shard-count 1
aws --endpoint-url=http://localstack:4566 kinesis wait stream-exists --stream-name ditto
```

**Comment by benfoster** [comment]
Windows issue?

**Comment by joe-hutchinson-cko** [comment]
Likely

File: docker-setup/localstack/init.sh (hunk 6 above)
Comment (joe-hutchinson-cko) [reply docker-setup/localstack/init.sh:1]: Very minor but you could use `awslocal`

File: src/Ditto.Kinesis/ReplicatingConsumer.cs (hunk 5 above)
Comment (ben-foster-cko) [reply src/Ditto.Kinesis/ReplicatingConsumer.cs:72]: As discussed, we can use the Standard retry mode by setting an environment variable in AWS

**Review by joe-hutchinson-cko — APPROVED** [review]
Only remaining improvement would be to batch write events to Kinesis, ideally dealing with a batch of messages from EVS so we only ack EVS once we've published to Kinesis.

**Pull request merged**